File: doc-examples/jdbc-example-python/src/main/python/example/FlightRepository.py
from java.lang import Void
from java.util.concurrent import CompletableFuture
from micronaut.data.annotation import Upsert
from micronaut.data.jdbc.annotation import JdbcRepository
from micronaut.data.model.query.builder.sql import Dialect
from micronaut.data.repository import CrudRepository

from example.Flight import Flight


# tag::upsert-repository[]
@JdbcRepository(dialect=Dialect.H2)
class FlightRepository(CrudRepository[Flight, str]):

    def upsert(self, flight: Flight) -> None: ...

    def upsertAll(self, flights: list[Flight]) -> None: ...

    @Upsert
    def put(self, flight: Flight) -> None: ...

    @Upsert
    def putAll(self, flights: list[Flight]) -> None: ...

    @Upsert
    def upsertFuture(self, flight: Flight) -> CompletableFuture[Void]: ...

    @Upsert
    def upsertAllFuture(self, flights: list[Flight]) -> CompletableFuture[Void]: ...
# end::upsert-repository[]

File: doc-examples/jdbc-example-java/src/main/java/example/FlightRepository.java
package example;

import io.micronaut.context.annotation.Requires;
import io.micronaut.data.annotation.Upsert;
import io.micronaut.data.jdbc.annotation.JdbcRepository;
import io.micronaut.data.model.query.builder.sql.Dialect;
import io.micronaut.data.repository.CrudRepository;

import java.util.concurrent.CompletableFuture;

@Requires(notEnv="oracle")
// tag::upsert-repository[]
@JdbcRepository(dialect = Dialect.H2)
public interface FlightRepository extends CrudRepository<Flight, String> {

    void upsert(Flight flight);

    void upsertAll(Iterable<Flight> flights);

    @Upsert
    void put(Flight flight);

    @Upsert
    void put(Iterable<Flight> flights);

    @Upsert
    CompletableFuture<Void> upsertFuture(Flight flight);

    @Upsert
    CompletableFuture<Void> upsertFuture(Iterable<Flight> flights);
}
// end::upsert-repository[]
